tools/eslint-plugin/dist/rules/no-internal-barrel-imports.js

```diff
@@ -0,0 +1,297 @@
+```
+
+fixing is currently restricted to named exports only.
+*/
+
+"use strict";
+
+const { getParserServices } = require("./utils/parser");
+const ts = require("typescript");
+/** for dealing with paths on the current os */
+const OsPaths = require("path");
+/** for dealing with paths in typescript import specifiers */
+const TsImportPaths = OsPaths.posix;
+
+const OPTION_IGNORED_BARREL_MODULES = "ignored-barrel-modules";
+
+const messages = {
+  noInternalBarrelImports: `Do not consume barrel imports within the same package`,
+};
+
+/** @type {typeof messages} */
+const messageIds = Object.keys(messages).reduce((obj, key) => {
+  obj[key] = key;
+  return obj;
+}, {});
+
+/** @type {import("eslint").Rule.RuleModule} */
+const rule = {
+  meta: {
+    type: "problem",
+    docs: {
+      description:
+        "Reports uses of re-exports within a package, in order to prevent such cyclic dependencies.",
+      category: "TypeScript",
+    },
+    schema: [
+      {
+        type: "object",
+        additionalProperties: false,
+        properties: {
+          /**
+           * This option may be replaced in the future with "required-barrel-modules", which
+           * indicates that all modules importing a symbol re-exported by that barrel module
+           * must import it from the barrel module, and from nowhere else. In some complicated dependency situations,
+           * forcing usage of a barrel module will enforce a particular dependency order. Modules that are imported by the barrel
+           * file will not need to import sibling modules from the barrel.
+           */
+          [OPTION_IGNORED_BARREL_MODULES]: {
+            type: "array",
+            description: "Usage of re-exports from these paths are ignored.",
+            default: [],
+          },
+        },
+      },
+    ],
+    fixable: "code",
+    messages,
+  },
+
+  create(context) {
+    const parserServices = getParserServices(context);
+    /** @type {ts.Program} */
+    const program = parserServices.program;
+    const checker = program.getTypeChecker();
+    const maybeTsConfig = program.getCompilerOptions().configFilePath;
+
+    const extraOpts = context.options[0];
+    const ignoredBarrelModules = (
+      (maybeTsConfig &&
+        extraOpts &&
+        extraOpts[OPTION_IGNORED_BARREL_MODULES]) ||
+      []
+    ).map((p) => {
+      /** @type {string} earlier check means it is definitely a string */
+      const tsConfig = maybeTsConfig;
+      return OsPaths.normalize(OsPaths.resolve(OsPaths.dirname(tsConfig), p));
+    });
+
+    return {
+      ImportDeclaration(node) {
+        /** @param {ts.Symbol | undefined} symbol */
+        function getRelativeImportPathForExportedSymbol(symbol) {
+          if (symbol === undefined) return "";
+          const declaration = symbol.valueDeclaration || symbol.declarations[0];
+          const fileOfExport = declaration.getSourceFile();
+          const path = TsImportPaths.normalize(
+            withoutExt(
```

Comment:
I'd move withoutExt and getRelativeImportPathForExportedSymbol closer together, but I'm ok with it either way